crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -146,31 +134,49 @@ impl SystemRegistry {
     ///
     /// Systems will be run sequentially in registration order if more than one registered system matches the provided set.
     #[inline]
-    pub fn run_callback(
+    pub fn run_system_by_id(
         &mut self,
         world: &mut World,
-        callback: Callback,
+        id: SystemId,
     ) -> Result<(), SystemRegistryError> {
-        let boxed_set = callback.set;
-        match self.systems.get_mut(&boxed_set) {
-            Some(matching_system) => {
+        match self.systems.get_mut(id.index()) {
+            Some((initialized, matching_system)) => {
+                if !*initialized {
+                    matching_system.initialize(world);
+                    *initialized = true;
+                }
                 matching_system.run((), world);
                 matching_system.apply_buffers(world);
                 Ok(())
             }
-            None => Err(SystemRegistryError::SetNotFound(boxed_set)),
+            None => Err(SystemRegistryError::SystemIdNotRegistered(id)),
         }
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub struct SystemId(usize);
```

Comment:
@alice-i-cecile I've introduced `SystemId`s to use them like `Components` use `ComponentId`s.

There are two main reasons for this:
- I find it confusing to use either `BoxedSystemSet` or `SystemTypeSet<dyn Any>` as keys to reference systems present in the registry. My intuition is to think of groups of systems whenever is read "set".
- This was the API can force you to preregister systems whenever you need to use a `SystemId`. I think it can cause confusion if `run_callback(Callback::new(my_system))` panics but `run_system(my_system)` works fine.

What do you think of this approach? (It still needs some love to improve system initialization and update the docs)

Reply:
I still don't like being forced to preregister systems: can't we make sure neither case panics?

I think we should stick to `SystemTypeSet<dyn Any>` as our key type, but add a friendlier type alias for it.

Reply:
Understood, I'll create the type alias

Reply:
Avoiding preregistration for callbacks is tricky. I've done some experiments with an `IntoBoxedSystem` trait so that we could store `Box<dyn IntoBoxedSytem<(), ()>` without specifying the type of the system. I don't now if something like `Box<IntoSystem<(), (), System = dyn System>>` could work with enough `?Sized`.

Reply:
Yeah that's definitely the largest challenge there.